File: skellytracker/trackers/base_tracker/point_cloud.py
"""
PointCloud: a labeled array of 3D points with visibility scores.

Names and coordinates are structurally coupled — they cannot desync.
This is the canonical data primitive for tracked landmarks throughout
the pipeline: detection → triangulation → filtering → visualization.
"""

from dataclasses import dataclass, field

import numpy as np
from numpy.typing import NDArray


@dataclass(slots=True)
class PointCloud:
    """
    A fixed-size labeled point array.

    Stores N named points, each with xyz coordinates and a visibility score.
    Names are immutable and set at construction — the i-th name always
    corresponds to the i-th row of the coordinate array.

    Points that are not detected have NaN coordinates and zero visibility.
    """

    names: tuple[str, ...]
    xyz: NDArray[np.floating]        # (N, 3) — mutable coordinates
    visibility: NDArray[np.floating] # (N,)   — mutable confidence scores
    _name_to_idx: dict[str, int] = field(init=False, repr=False)

    def __post_init__(self) -> None:
        n = len(self.names)
        if self.xyz.shape != (n, 3):
            raise ValueError(f"xyz shape {self.xyz.shape} does not match {n} names (expected ({n}, 3))")
        if self.visibility.shape != (n,):
            raise ValueError(f"visibility shape {self.visibility.shape} does not match {n} names (expected ({n},))")
        self._name_to_idx = {name: i for i, name in enumerate(self.names)}

    # =========================================================================
    # Properties — zero-copy views
    # =========================================================================

    @property
    def n_points(self) -> int:
        return len(self.names)

    @property
    def xy(self) -> NDArray[np.floating]:
        """(N, 2) view of just x, y coordinates. Zero-copy."""
        return self.xyz[:, :2]

    @property
    def valid_mask(self) -> NDArray[np.bool_]:
        """Boolean mask: True where point has non-NaN coordinates."""
        return ~np.isnan(self.xyz).any(axis=1)

    @property
    def n_valid(self) -> int:
        return int(self.valid_mask.sum())

    # =========================================================================
    # Lookup
    # =========================================================================

    def index_of(self, name: str) -> int:
        """Get the row index for a point name. Raises KeyError if not found."""
        return self._name_to_idx[name]

    def has_name(self, name: str) -> bool:
        return name in self._name_to_idx

    def xyz_by_name(self, name: str) -> NDArray[np.floating]:
        """Get (3,) coordinate array for a single point by name."""
        return self.xyz[self._name_to_idx[name]]

    def xy_by_name(self, name: str) -> NDArray[np.floating]:
        """Get (2,) coordinate array for a single point by name."""
        return self.xyz[self._name_to_idx[name], :2]

    # =========================================================================
    # Slicing — returns views, not copies
    # =========================================================================

    def slice(self, start: int, stop: int) -> "PointCloud":
        """
        Return a PointCloud backed by views into this cloud's arrays.

        The returned cloud shares memory with this one — mutations
        to one affect the other.
        """
        return PointCloud(
            names=self.names[start:stop],
            xyz=self.xyz[start:stop],
            visibility=self.visibility[start:stop],
        )

    def slice_by_names(self, names: tuple[str, ...]) -> "PointCloud":
        """Return a PointCloud with only the named points (copies data)."""
        indices = np.array([self._name_to_idx[n] for n in names])
        return PointCloud(
            names=names,
            xyz=self.xyz[indices].copy(),
            visibility=self.visibility[indices].copy(),
        )

    # =========================================================================
    # Construction helpers
    # =========================================================================

    @staticmethod
    def empty(names: tuple[str, ...]) -> "PointCloud":
        """Create a PointCloud with all NaN coordinates and zero visibility."""
        n = len(names)
        return PointCloud(
            names=names,
            xyz=np.full((n, 3), np.nan),
            visibility=np.zeros(n),
        )

    @staticmethod
    def concatenate(clouds: list["PointCloud"]) -> "PointCloud":
        """
        Concatenate multiple PointClouds into one.

        Copies data — the result is independent of the inputs.
        """
        if not clouds:
            raise ValueError("Cannot concatenate empty list of PointClouds")

        all_names: list[str] = []
        all_xyz: list[NDArray] = []
        all_vis: list[NDArray] = []

        for cloud in clouds:
            all_names.extend(cloud.names)
            all_xyz.append(cloud.xyz)
            all_vis.append(cloud.visibility)

        return PointCloud(
            names=tuple(all_names),
            xyz=np.concatenate(all_xyz, axis=0).copy(),
            visibility=np.concatenate(all_vis, axis=0).copy(),
        )

    # =========================================================================
    # Conversion — for interfaces that expect dicts or flat arrays
    # =========================================================================

    def to_2d_array(self) -> NDArray[np.floating]:
        """
        Return (N, 2) array of xy coordinates.

        Always returns N rows in name order. NaN for undetected points.
        Copy — safe to mutate without affecting the cloud.
        """
        return self.xy.copy()

    def to_named_dict(self, dimensions: int = 2) -> dict[str, NDArray[np.floating]]:
        """
        Return dict of {name: coordinate_array} for ALL points.

        Always includes all N names, preserving structural consistency
        with to_2d_array(). NaN-valued points are included.
        """
        d = dimensions
        return {name: self.xyz[i, :d].copy() for i, name in enumerate(self.names)}

    def to_valid_dict(self, dimensions: int = 2) -> dict[str, NDArray[np.floating]]:
        """
        Return dict of {name: coordinate_array} for only valid (non-NaN) points.

        Use this for display/overlay purposes where you want to skip missing points.
        Do NOT use this for triangulation — use to_named_dict() instead.
        """
        mask = self.valid_mask
        d = dimensions
        return {
            name: self.xyz[i, :d].copy()
            for i, name in enumerate(self.names)
            if mask[i]
        }

    def to_scaled_tuples(self, dimensions: int, scale_by: float = 1.0) -> dict[str, tuple]:
        """
        Return dict of {name: (x, y[, z])} for only valid points, scaled.

        Matches the legacy MediapipeObservation.all_points() interface.
        """
        mask = self.valid_mask
        d = dimensions
        scaled = self.xyz * scale_by
        return {
            name: tuple(scaled[i, :d])
            for i, name in enumerate(self.names)
            if mask[i]
        }

    # =========================================================================
    # Filtering
    # =========================================================================

    def filtered_by_confidence(self, threshold: float, fill_with_nans: bool = True) -> "PointCloud":
        """
        Return a new PointCloud with low-confidence points masked.

        If fill_with_nans is True, returns same-sized cloud with NaN for
        low-confidence points. If False, returns a smaller cloud with
        only high-confidence points (names change!).
        """
        if fill_with_nans:
            xyz = self.xyz.copy()
            vis = self.visibility.copy()
            mask = vis < threshold
            xyz[mask] = np.nan
            return PointCloud(names=self.names, xyz=xyz, visibility=vis)
        else:
            mask = self.visibility >= threshold
            indices = np.where(mask)[0]
            return PointCloud(
                names=tuple(self.names[i] for i in indices),
                xyz=self.xyz[indices].copy(),
                visibility=self.visibility[indices].copy(),
            )

    # =========================================================================
    # Repr
    # =========================================================================

    def __len__(self) -> int:
        return self.n_points

File: skellytracker-rust/src/core/point_cloud.rs
use std::collections::HashMap;
use ndarray::{Array1, Array2};

/// Canonical data primitive for tracked landmarks.
///
/// The i-th name in `names` corresponds to the i-th row in `xyz`
/// and the i-th element in `visibility`.
#[derive(Debug, Clone)]
pub struct PointCloud {
    pub names: Vec<String>,
    pub xyz: Array2<f64>,             // (N, 3), z=0 for 2D-only trackers
    pub visibility: Array1<f64>,       // (N,), 0.0 = not detected
    name_to_index: HashMap<String, usize>,
}

impl PointCloud {
    /// Create a new PointCloud with all-NaN coordinates and zero visibility.
    pub fn new(names: Vec<String>) -> Self {
        let number_of_points = names.len();
        let mut name_to_index = HashMap::with_capacity(number_of_points);
        for (index, name) in names.iter().enumerate() {
            name_to_index.insert(name.clone(), index);
        }
        Self {
            names,
            xyz: Array2::from_elem((number_of_points, 3), f64::NAN),
            visibility: Array1::zeros(number_of_points),
            name_to_index,
        }
    }

    /// Look up the array index for a named point. O(1).
    pub fn index_of(&self, name: &str) -> Option<usize> {
        self.name_to_index.get(name).copied()
    }

    /// Get the (x, y, z) coordinates for a named point.
    pub fn coordinates_by_name(&self, name: &str) -> Option<[f64; 3]> {
        let index = self.name_to_index.get(name)?;
        let row = self.xyz.row(*index);
        Some([row[0], row[1], row[2]])
    }

    /// Immutable view of the xy columns. Zero-copy.
    pub fn xy_view(&self) -> ndarray::ArrayView2<f64> {
        self.xyz.slice(ndarray::s![.., 0..2])
    }

    /// Number of points in this cloud.
    pub fn number_of_points(&self) -> usize {
        self.names.len()
    }

    /// Mask of points with valid (non-NaN) coordinates.
    pub fn valid_mask(&self) -> Array1<bool> {
        self.xyz.column(0).mapv(|x| !x.is_nan())
    }

    /// Count of points with valid (non-NaN) coordinates.
    pub fn number_of_valid(&self) -> usize {
        self.valid_mask().iter().filter(|&&valid| valid).count()
    }

    /// Copy out a subset of points by name. Returns a new PointCloud.
    pub fn slice_by_names(&self, requested_names: &[&str]) -> Self {
        let mut new_cloud = Self::new(
            requested_names.iter().map(|n| n.to_string()).collect(),
        );
        for (new_index, name) in requested_names.iter().enumerate() {
            if let Some(old_index) = self.name_to_index.get(*name) {
                let row = self.xyz.row(*old_index);
                new_cloud.xyz.row_mut(new_index).assign(&row);
                new_cloud.visibility[new_index] = self.visibility[*old_index];
            }
        }
        new_cloud
    }

    /// Filter points by confidence threshold. Returns a new PointCloud.
    /// If `fill_with_nans` is true, points below threshold get NaN coordinates.
    pub fn filtered_by_confidence(
        &self,
        threshold: f64,
        fill_with_nans: bool,
    ) -> Self {
        let mut filtered = self.clone();
        for index in 0..self.number_of_points() {
            if self.visibility[index] < threshold {
                filtered.visibility[index] = 0.0;
                if fill_with_nans {
                    filtered.xyz.row_mut(index).fill(f64::NAN);
                }
            }
        }
        filtered
    }

    /// Convert to a (N, 2) array of xy coordinates (always N rows, NaN fill).
    pub fn to_2d_array(&self) -> Array2<f64> {
        self.xy_view().to_owned()
    }
}
